helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -122,11 +222,72 @@ public void close() {
     if (isCurrentOwner()) {
       throw new HelixException("Please unlock the lock before closing it.");
     }
+    _baseDataAccessor.unsubscribeDataChanges(_lockPath, this);
     _baseDataAccessor.close();
   }
 
+  @Override
+  public void handleDataChange(String dataPath, Object data) throws Exception {
+    Stat stat = new Stat();
+    ZNRecord readData =
+        _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+    LockInfo lockInfo = new LockInfo(readData);
+    // We are the current owner
+    if (isCurrentOwner(lockInfo)) {
+      if (lockInfo.getRequestorId().equals("NONE") || lockInfo.getPriority() > lockInfo
+          .getRequestorPriority()) {
+        LOG.debug("We do not need to handle this data change");
+      } else {
+        _lockListener.onCleanupNotification();
+        CleanupUpdater cleanupUpdater = new CleanupUpdater();
+        boolean res = _baseDataAccessor.update(_lockPath, cleanupUpdater, AccessOption.PERSISTENT);
+        if (!res) {
+          throw new HelixException(
+              String.format("User %s failed to update lock path %s", _userId, _lockPath));
+        }
+      }
+    } // We are the current requestor
+    else if (lockInfo.getRequestorId().equals(_userId)) {
+      // In case the owner field is empty, it means previous owner has finished cleanup work.
+      if (lockInfo.getOwner().equals("NONE")) {
+        ZNRecord znRecord = composeNewOwnerRecord();
+        LockUpdater updater = new LockUpdater(new LockInfo(znRecord));
+        _baseDataAccessor.update(_lockPath, updater, AccessOption.PERSISTENT);
+        onAcquiredLockNotification();
+      } else {
+        LOG.info("We do not need to handle this data change");
+      }
+    } // If we are waiting for the lock, but find we are not the requestor any more, meaning we
+    // are preempted by an even higher priority request
+    else if (!lockInfo.getRequestorId().equals("NONE") && !lockInfo.getRequestorId().equals(_userId)
+        && _lockStatus == LockConstants.LockStatus.PENDING) {
+      onDeniedPendingLockNotification();
+    }
+  }
+
+  /**
+   * call back called when the lock is acquired
+   */
+  public void onAcquiredLockNotification() {
+    _lockStatus = LockConstants.LockStatus.LOCKED;
+    _countDownLatch.countDown();
+  }
+
   /**
-   * Class that specifies how a lock node should be updated with another lock node
+   * call back called when the pending request is denied due to another higher priority request
+   */
+  public void onDeniedPendingLockNotification() {
+    _lockStatus = LockConstants.LockStatus.PREEMPTED;
+    _countDownLatch.countDown();
+  }
+
+  @Override
+  public void handleDataDeleted(String dataPath) throws Exception {
+  }
```

Comment:
Once we support LOCK node cleanup, then we need to add some logic here, right?
Maybe some TODO comments would be helpful.